lib/plugin-loader.ts

```diff
@@ -0,0 +1,853 @@
+    self.loadTranslations();
+    return self;
+  }
+}
+
+export class PluginLoader {
+  private plugins: any;
+  private pluginMap: any;
+  public intervalScanner: any;
+  public makePlugin: any;
+
+  constructor(private options: any) {
+    EventEmitter.call(this);
+    this.options = zluxUtil.makeOptionsObject(defaultOptions, options);
+    this.plugins = [];
+    this.pluginMap = {};
+  }
+
+  private _readPluginDef(pluginDescriptorFilename: string) {
+    const pluginPtrPath = this.options.relativePathResolver(pluginDescriptorFilename,
+                                                 this.options.pluginsDir);
+    bootstrapLogger.info(`ZWED0044I`, pluginPtrPath); //bootstrapLogger.info(`Processing plugin reference ${pluginPtrPath}...`);
+    if (!fs.existsSync(pluginPtrPath)) {
+      throw new Error(`ZWED0021E - ${pluginPtrPath} is missing`);
+    }
+    const pluginPtrDef = jsonUtils.parseJSONWithComments(pluginPtrPath);
+    bootstrapLogger.debug("ZWED0121I", util.inspect(pluginPtrDef)); //bootstrapLogger.log(bootstrapLogger.FINER, util.inspect(pluginPtrDef));
+    let pluginBasePath = pluginPtrDef.pluginLocation;
+    if (!path.isAbsolute(pluginBasePath)) {
+      pluginBasePath = this.options.relativePathResolver(pluginBasePath, process.cwd());
+    }
+    if (!fs.existsSync(pluginBasePath)) {
+      return {location: pluginBasePath,
+              identifier: pluginPtrDef.identifier,
+              error: new Error(`ZWED0015E - No plugin directory found at ${pluginPtrDef.pluginLocation}`)};
+     }
+    let pluginDefPath = path.join(pluginBasePath, 'pluginDefinition.json');
+    if (!fs.existsSync(pluginDefPath)) {
+      return {location: pluginBasePath,
+              identifier: pluginPtrDef.identifier,
+              error: new Error(`ZWED0016E - No pluginDefinition.json found at ${pluginBasePath}`)};
+    }
+    let pluginDef = jsonUtils.parseJSONWithComments(pluginDefPath);
+    bootstrapLogger.debug("ZWED0122I", util.inspect(pluginDef)); //bootstrapLogger.log(bootstrapLogger.FINER,util.inspect(pluginDef));
+    if (pluginDef.identifier !== pluginPtrDef.identifier) {
+      return {location: pluginBasePath,
+              identifier: pluginPtrDef.identifier,
+              error: new Error(`ZWED0017E - Identifier doesn't match one found in pluginDefinition: ${pluginDef.identifier}`)};
+    }
+    if (!pluginDef.pluginType) {
+      return {location: pluginBasePath,
+              identifier: pluginPtrDef.identifier,
+              error: new Error(`ZWED0018E - No plugin type found, skipping`)};
+    }
+    bootstrapLogger.info(`ZWED0214I`, pluginBasePath, pluginDef.identifier, pluginDef.pluginType); //bootstrapLogger.info(`Read ${pluginBasePath}: found plugin id = ${pluginDef.identifier}, `
+        //+ `type = ${pluginDef.pluginType}`);
+    pluginDef.location = pluginBasePath;
+    return pluginDef;
+  }
+
+  private _readPluginDefAsync(pluginDescriptorFilename: any) {
+    return new Promise((resolve, reject)=> {
```

Comment:
It's especially important to show the return type when the return type is a promise.
`private _readPluginDefAsync(pluginDescriptorFilename: any): Promise<any>` or `private _readPluginDefAsync(pluginDescriptorFilename: any): Promise<object>`  seems correct here.